pennylane_lightning/src/gates/cpu_kernels/avx_common/ApplyCRY.hpp

```diff
@@ -0,0 +1,323 @@
+    static auto applyInternalInternalDiagFactor(ParamT angle) {
+        std::array<PrecisionT, packed_size> arr{};
+
+        // positions are after permutations
+        for (size_t k = 0; k < packed_size / 2; k++) {
+            if ((k >> control) & 1U) { // if control bit is 1
+                arr[2 * k + 0] = std::cos(angle / 2);
+                arr[2 * k + 1] = std::cos(angle / 2);
+            } else {
+                arr[2 * k + 0] = Precision{1.0};
+                arr[2 * k + 1] = Precision{1.0};
+            }
+        }
+        return set<Precision, packed_size>(arr);
+    }
+
+    template <size_t control, size_t target, class ParamT>
+    static void applyInternalInternal(std::complex<PrecisionT> *arr,
+                                      size_t num_qubits, bool inverse,
+                                      ParamT angle) {
+        constexpr static auto perm =
+            applyInternalInternalPermutation<control, target>();
+
+        if (inverse) {
+            angle *= -1.0;
+        }
+
+        const auto off_diag_factor =
+            applyInternalInternalOffDiagFactor<control, target>(angle);
+        const auto diag_factor =
+            applyInternalInternalDiagFactor<control, target>(angle);
+
+        for (size_t n = 0; n < exp2(num_qubits); n += packed_size / 2) {
+            const auto v = PrecisionAVXConcept::load(arr + n);
+            const auto diag_w = diag_factor * v;
+            const auto off_diag_w =
+                off_diag_factor * Permutation::permute<perm>(v);
+            PrecisionAVXConcept::store(arr + n, diag_w + off_diag_w);
+        }
+    }
+
+    /**
+     * @brief Factor for (1).
+     * [1, 1, 1, 1, cos(phi/2), cos(phi/2), cos(phi/2), cos(phi/2)]
+     */
+    template <size_t control, typename ParamT>
+    static auto applyInternalExternalDiagFactor(ParamT angle) {
+        std::array<Precision, packed_size> arr{};
+
+        for (size_t k = 0; k < packed_size / 2; k++) {
+            if ((k >> control) & 1U) {
+                // if control is 1
+                arr[2 * k + 0] = std::cos(angle / 2);
+                arr[2 * k + 1] = std::cos(angle / 2);
+            } else {
+                arr[2 * k + 0] = 1.0;
+                arr[2 * k + 1] = 1.0;
+            }
+        }
+        return set(arr);
+    }
+
+    /**
+     * @brief Factor for (2) when the target bit is 1
+     */
+    template <size_t control, typename ParamT>
+    static auto applyInternalExternalOffDiagFactor(ParamT angle) {
+        std::array<Precision, packed_size> arr{};
+
+        for (size_t k = 0; k < packed_size / 2; k++) {
+            if ((k >> control) & 1U) {
+                // if control is 1
+                arr[2 * k + 0] = std::sin(angle / 2);
+                arr[2 * k + 1] = std::sin(angle / 2);
+            } else {
+                arr[2 * k + 0] = 0.0;
+                arr[2 * k + 1] = 0.0;
+            }
+        }
+        return set(arr);
```

Comment:
Still unsure about `set` here and throughout, but maybe I am missing something here.